### PR TITLE
fix(gatsby): always add both `childField` and `childrenField` in GraphQL

### PR DESCRIPTION
**Pull request by vladar**
## Description

Before this PR we create only one of the child fields depending on the number of elements in the node `children` array.

E.g. when `children` array contains 1 item we create this field:

```graphql
type Foo {
  childBar: Bar
}
```

But when `children` array contains 2+ items, we create this field:

```graphql
type Foo {
  childrenBar: [Bar]
}
```

The problem is that it relies on inference. And if at some point in time you had a single child node and then suddenly added a second child node - your schema will change from `childBar` to `childrenBar` and your queries will break.

With this PR we always add both fields instead:

```graphql
type Foo {
  childBar: Bar
  childrenBar: [Bar]
}
```

In this case, `childBar` will always return the first child from the array of node children. And then it is up to users to pick the most appropriate field for their queries.

This also deprecates the `many` argument of the `childOf` directive [^1]

This is a non-breaking change. Also required to unblock https://github.com/gatsbyjs/gatsby/pull/28483

Credits for this idea go to @pieh 

### Documentation

- [childOf directive][^1]
- [Child/parent fields][^2]
- [Schema generation][^3]

[^1]: https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#defining-child-relations
[^2]: https://www.gatsbyjs.com/docs/schema-inference/#childparent-fields
[^3]: https://www.gatsbyjs.com/docs/schema-generation/#4-parent--children-relationships

## Related Issues

https://github.com/gatsbyjs/gatsby/pull/28483

### CODE REVIEW

File: packages/gatsby/src/schema/schema.js [hunk 1]
```diff
@@ -968,12 +969,9 @@ const addConvenienceChildrenFields = ({ schemaComposer }) => {
...
     }
-    children.forEach((many, child) => {
-      if (many) {
-        typeComposer.addFields(createChildrenField(child.getTypeName()))
-      } else {
-        typeComposer.addFields(createChildField(child.getTypeName()))
-      }
+    children.forEach(child => {
+      typeComposer.addFields(createChildrenField(child.getTypeName()))
+      typeComposer.addFields(createChildField(child.getTypeName()))
```
Comment (vladar) [inline packages/gatsby/src/schema/schema.js:974]: That's basically the essence of this PR. Other changed files are just tests/docs updates.

File: packages/gatsby/src/schema/schema.js [hunk 2]
```diff
@@ -992,12 +990,9 @@ const addConvenienceChildrenFields = ({ schemaComposer }) => {
...
         }
-        children.forEach((many, child) => {
-          if (many) {
-            typeComposer.addFields(createChildrenField(child.getTypeName()))
-          } else {
-            typeComposer.addFields(createChildField(child.getTypeName()))
-          }
+        children.forEach(child => {
+          typeComposer.addFields(createChildrenField(child.getTypeName()))
+          typeComposer.addFields(createChildField(child.getTypeName()))
```
Comment (vladar) [inline packages/gatsby/src/schema/schema.js:995]: Same change, just in the different codepath

File: packages/gatsby/src/schema/schema.js [hunk 3]
```diff
@@ -1021,46 +1016,39 @@ const addImplicitConvenienceChildrenFields = ({
...
       }
     }
 
-    if (maxChildCount > 1) {
-      typeComposer.addFields(createChildrenField(typeName))
-    } else {
-      typeComposer.addFields(createChildField(typeName))
-    }
+    typeComposer.addFields(createChildrenField(typeName))
+    typeComposer.addFields(createChildField(typeName))
```
Comment (vladar) [inline packages/gatsby/src/schema/schema.js:1051]: And one more entry. Other changes are mostly chores and don't have any effect on behavior.

**Comment by pieh** [comment]
Just to leave papar trail for the future from our offline chat:
> do you think, given that we will add both of them, it would help a little to add description to those fields? https://github.com/gatsbyjs/gatsby/blob/2bb4214d36616ae593a8a82d63102dca2d5e44e6/packages/gatsby/src/schema/schema.js#L1067-L1100
> another question is wether we should split PR due to docs site building from master branch for now 
similar case to ( https://github.com/gatsbyjs/gatsby/pull/28666 )

**Review by pieh — APPROVED** [review]
Thank you!

**Pull request merged**